File: python/src/selftune.py
"""
This module implements SelfTune, a framework that uses data-driven tuning of
parameters that produce larger rewards.
"""
import math
from typing import List, Union, Tuple, Optional
import numpy as np


class Constraints:
    """A class to hold constraints on a parameter learned using SelfTune.

    Attributes:
        min: The minimum value a parameter can take.
        max: The maximum value a parameter can take.
        is_int: Specifies whether a parameter is an integer.
    """
    DEFAULT_MIN = 2.2250738585072014E-308  # 64 bit floating point min value
    DEFAULT_MAX = 1.7976931348623158E+308  # 64 bit floating point max value

    def __init__(self,
                 c_min: Union[int, float] = DEFAULT_MIN,
                 c_max: Union[int, float] = DEFAULT_MAX,
                 is_int: bool = False):
        self.min = c_min
        self.max = c_max
        self.is_int = is_int


class TaskData:
    """A helper class that holds data representing the learning task.

    Attributes:
        initial_values: Initial values for the parameters.
        constraints: Constraints on the parameters.
        opt: The optimization algorithm to use. Can take values {"bluefin"}.
        feedback: Type of feedback update. Can take values
            {"onepoint", "twopoint"}.
        eta: The value of the learning rate in the ongoing round.
        initial_eta: The initial value of eta.
        delta: The exploration radius.
        eta_decay_rate: The decay rate of eta.
        is_ints: A numpy array of the is_int constraint for all parameters.
            Used to optimize computations.
        feedback_denom: Used in the gradient estimation process. The value is
            1 for onepoint and 2 for twopoint.
    """

    def __init__(self, initial_values: np.ndarray,
                 constraints: Optional[List[Constraints]], opt: str,
                 feedback: str, eta: float, delta: float,
                 eta_decay_rate: float):
        self.initial_values = initial_values
        self.constraints = constraints
        self.opt = opt
        self.feedback = feedback
        self.eta = eta
        self.initial_eta = eta
        self.delta = delta
        self.eta_decay_rate = eta_decay_rate
        self.is_ints = np.asarray([i.is_int for i in constraints])

        # Used during gradient estimation
        self.feedback_denom = 1 if feedback == 'onepoint' else 2

    def validate_params(self):
        """Validates the task data."""
        n_parameters = self.initial_values.size

        assert len(self.initial_values.shape) == 1

        if self.constraints is not None:
            assert len(self.constraints) == n_parameters, \
                'Constraints not provided for all parameters'

        for constraint in self.constraints:
            assert constraint.min < constraint.max, \
                f'Invalid constraint: Min={constraint.min} >= ' \
                f'Max={constraint.max}'

        assert self.delta > 0, f'Delta={self.delta} should be > 0'


class SessionData:
    """A helper class that holds data representing the session.

    Attributes:
        first_explore: The first point to explore as part of this session.
        second_explore: The second point to explore as part of this session.
        center: The point around which exploration is done.
        explore_direction: Direction along which exploration is done.
        first_explore_id: Identifier for first_explore.
        second_explore_id: Identifier for second_explore.
        first_reward: Used to store the reward from the first explore during
            two-point feedback.
    """

    def __init__(self, center: np.ndarray, explore_direction: np.ndarray,
                 taskdata: TaskData, first_id: int):
        # Explore in the direction of explore_direction
        self.first_explore = SelfTune.compute_explore_value(
            center, 1, explore_direction, taskdata)
        # Explore in the direction opposite to explore_direction
        self.second_explore = None if taskdata.feedback == 'onepoint' else \
            SelfTune.compute_explore_value(
                center, -1, explore_direction, taskdata)
        self.center = center
        self.explore_direction = explore_direction
        self.first_explore_id = first_id
        self.second_explore_id = 0 if taskdata.feedback == 'onepoint' else first_id + 1
        self.first_reward = None


class SelfTune:
    """SelfTune Class.

    Attributes:
        initial_values: Initial values for the parameters.
        constraints: Constraints on the parameters.
        opt: The optimization algorithm to use. Can take values {'bluefin'}.
        feedback: Type of feedback update. Can take values
            {'onepoint', 'twopoint'}.
        eta: The learning rate.
        delta: The exploration radius.
        random_state: The random seed used to initialize the numpy
            pseudo-random number generator.
        eta_decay_rate: The decay rate of eta. Defaults to 0 where eta
            does not decay after each round.
        taskdata: Stores data relating to the learning task.
        rs: The random number generator used by the algorithm.
    """

    def __init__(self,
                 initial_values: np.ndarray,
                 constraints: Optional[List[Constraints]] = None,
                 opt: str = 'bluefin',
                 feedback: str = 'onepoint',
                 eta: float = 0.01,
                 delta: float = 0.1,
                 random_state: int = None,
                 eta_decay_rate: float = 0):
        self.round = 0

        self.taskdata = TaskData(initial_values, constraints, opt, feedback,
                                 eta, delta, eta_decay_rate)
        self.taskdata.validate_params()

        self.random_state = np.random.RandomState(seed=random_state)

        if self.taskdata.constraints is not None:
            initial_values = SelfTune.project(initial_values,
                                              self.taskdata.constraints,
                                              self.taskdata.delta)

        self.session = SessionData(
            initial_values, self.sample_unit_sphere(initial_values.shape),
            self.taskdata, 0)

    def predict(self, features=None) -> np.ndarray:
        """Gets the next set of parameter values to use.

        Args:
            features: features to use for the prediction. The bluefin
                algorithm does not use any features.

        Returns:
            A numpy array of the parameter values to use.
        """
        if self.session.first_reward is None:
            return self.session.first_explore

        return self.session.second_explore

    def set_reward(self, reward: float):
        """
        Sets the reward based on the parameters returned from predict.

        Args:
            reward: A measure of how well the parameter values returned in
                predict performed.
        """
        if self.taskdata.feedback_denom == 1:  # onepoint
            new_center = SelfTune.compute_new_center(
                self.session.center, self.session.explore_direction, reward,
                self.taskdata)
            self.session = SessionData(
                new_center, self.sample_unit_sphere(new_center.shape),
                self.taskdata, self.session.first_explore + 1)
            self.round += 1
            self.taskdata.eta = self.eta_decay()  # Eta Decay

        else:  # twopoint
            if self.session.first_reward is None:
                self.session.first_reward = reward
            else:
                reward_diff = (self.session.first_reward - reward)

                # Normalizing reward diff
                reward_diff /= max(abs(self.session.first_reward), abs(reward))

                new_center = SelfTune.compute_new_center(
                    self.session.center, self.session.explore_direction,
                    reward_diff, self.taskdata)
                self.session = SessionData(
                    new_center, self.sample_unit_sphere(new_center.shape),
                    self.taskdata, self.session.second_explore + 1)
                self.round += 1
                self.taskdata.eta = self.eta_decay()  # Eta Decay

    def sample_unit_sphere(self, shape: Tuple) -> np.ndarray:
        """Sample uniformly from the surface of a sphere.

        Args:
            shape: The shape of the returned sample.

        Returns:
            A vector with values sampled from the surface of a sphere in shape
            dimensions.
        """
        # Sample from a standard normal distribution i.e ~ N(0, 1).
        sample = self.random_state.normal(size=shape, loc=0, scale=1)
        norm = np.linalg.norm(sample)

        if norm == 0:  # norm == 0 is very unlikely, nevertheless handling it
            sample[0] = self.random_state.randint(0, 2) * 2 - 1
        else:
            sample /= np.linalg.norm(sample)

        # print(f'sample.Add(new List<double>{{{sample[0]}, {sample[1]}}});')
        return sample

    def eta_decay(self):
        """Decay eta based on the decay rate and current round.

        Returns:
            The new value of eta.
        """
        return self.taskdata.initial_eta / (
            1 + self.taskdata.eta_decay_rate * self.round)

    @property
    def center(self) -> np.ndarray:
        return self.session.center

    @property
    def predict_id(self) -> int:
        return self.session.first_explore_id if self.session.first_reward is None else self.session.second_explore_id

    @property
    def eta(self) -> float:
        return self.taskdata.eta

    @eta.setter
    def eta(self, e):
        assert e > 0, f'Eta={e} should be > 0'
        self.taskdata.eta = e

    @property
    def delta(self) -> float:
        return self.taskdata.delta

    @delta.setter
    def delta(self, d):
        assert d > 0, f'Delta={d} should be > 0'
        self.taskdata.delta = d

    @staticmethod
    def compute_new_center(center: np.ndarray, explore_direction: np.ndarray,
                           reward_grad: float,
                           taskdata: TaskData) -> np.ndarray:
        """Computes the updated set of parameters.

        Computes the updated set of parameters using the reward(s) obtained
        during the exploration around the current set of parameters.

        Args:
            center: The current parameter vector.
            explore_direction: Direction of exploration.
            reward_grad: Contribution of the reward(s) toward the gradient.
            taskdata: Task for which new center is to be computed.

        Returns:
            New set of parameters.
        """
        new_center = SelfTune.project(
            center +
            (taskdata.eta * center.size * reward_grad * explore_direction) /
            (taskdata.feedback_denom * taskdata.delta), taskdata.constraints,
            taskdata.delta)

        return new_center

    @staticmethod
    def compute_explore_value(center: np.ndarray, explore_sign: int,
                              explore_direction: np.ndarray,
                              taskdata: TaskData) -> np.ndarray:
        """Compute the point to explore.

        Computes the point to explore by randomly perturbing the center
        vector. The explore vector is sampled from a hypersphere centered at
        the center vector with a radius of delta.

        Args:
            center: The point around which exploration is done.
            explore_sign: Specifies whether exploration is in the direction of
                explore_dir or opposite to it.
            explore_direction: Direction of exploration.
            taskdata: Task for which the point to explore is computed.

        Returns:
            A numpy array that represents the point to explore.
        """
        change = explore_sign * taskdata.delta * explore_direction

        if taskdata.is_ints.any():
            rounded = np.round(change)
            change = np.where((rounded != 0) & (taskdata.is_ints), rounded,
                              change)
            change = np.where(
                (rounded == 0) & (change >= 0) & (taskdata.is_ints), 1, change)
            change = np.where(
                (rounded == 0) & (change < 0) & (taskdata.is_ints), -1, change)

        explore_value = center + change
        return explore_value

    @staticmethod
    def project(center: np.ndarray, constraints: List[Constraints],
                delta: float) -> np.ndarray:
        """Projects center vector based on constraints.

        Args:
            center: The vector of parameters to project.
            constraints: Constraints on the parameters.
            delta: The exploration radius of the learning task.


        Returns:
            The projected vector.
        """
        new_center = []
        for idx, val in enumerate(center):
            new_center.append(
                SelfTune.project_single(val, constraints[idx], delta))

        return np.array(new_center)

    @staticmethod
    def project_single(val: float, constraint: Constraints,
                       delta: float) -> float:
        """
        Projects a single value based on constraints.

        Args:
            val: The value to project.
            constraint: Constraints on the value.
            delta: The exploration radius of the learning task.

        Returns:
            The projected value.
        """
        is_int = constraint.is_int
        c_min = constraint.min
        c_max = constraint.max
        if is_int:
            c_min = math.ceil(c_min)
            c_max = math.floor(c_max)
            delta = math.ceil(delta)

        c_min += delta
        c_max -= delta

        new_val = min(max(c_min, val), c_max)
        if is_int:
            new_val = np.round(new_val)

        return new_val

File: c#/SelfTune/SelfTune.cs
// ---------------------------------------------------------------------------
// <copyright file="SelfTune.cs" company="Microsoft">
//     Copyright (c) Microsoft Corporation.  All rights reserved.
// </copyright>
// ---------------------------------------------------------------------------

namespace Microsoft.Research.SelfTune
{
    using System;
    using System.Collections.Generic;

    /// <summary>
    /// Type of feedback update.
    /// </summary>
    public enum Feedback : int
    {
        OnePoint = 1,
        TwoPoint = 2
    }

    /// <summary>
    /// Data-driven tuning of parameters towards values that produce larger rewards.
    /// </summary>
    public class SelfTune
    {
        private readonly TaskData taskData;
        private SessionData sessionData;

        private int round;

        /// <summary>
        /// The random number generator used by the algorithm.
        /// </summary>
        public Random random;

        /// <summary>
        /// A unique identifier for the point being returned in <see cref="Predict"/>.
        /// Changes after <see cref="SetReward(double)"/> (because a new point is returned
        /// then).
        /// </summary>
        public uint PredictId
        {
            get
            {
                return this.sessionData.FirstReward == null ? this.sessionData.FirstExploreId : this.sessionData.SecondExploreId;
            }
        }

        /// <summary>
        /// The point around which exploration is done.
        ///
        /// Note: This property provides some behind-the-scenes information for additional
        /// insight. Its existence can be safely ignored by the consumer of <see cref="SelfTune"/>.
        /// </summary>
        public IList<double> Center
        {
            get
            {
                int numParameters = this.taskData.InitialValues.Count;

                List<double> center = new List<double>(numParameters);

                for (int i = 0; i < numParameters; i++)
                {
                    center.Add(this.sessionData.Center[i]);
                }

                return center;
            }
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="initialValues">Initial values for the parameters.</param>
        /// <param name="constraints">Constraints on the parameters.</param>
        /// <param name="feedback">Type of feedback update to use.</param>
        /// <param name="eta">Hyperparameter for the algorithm - learning rate.</param>
        /// <param name="delta">Hyperparameter for the algorithm - exploration radius.</param>
        /// <param name="randomState">The random seed used to initialize the pseudo-random number generator.</param>
        /// <param name="etaDecayRate">The decay rate of eta. Defaults to 0 where eta does not decay after each round.</param>
        /// <exception cref="ArgumentException">When task related data isn't valid.</exception>
        public SelfTune(IList<double> initialValues, IList<Constraints> constraints, Feedback feedback, double eta, double delta, int? randomState = null, double etaDecayRate = 0.0)
        {
            TaskData taskData = new TaskData()
            {
                InitialValues = initialValues,
                Constraints = constraints,
                Feedback = feedback,
                Eta = eta,
                InitialEta = eta,
                Delta = delta,
                EtaDecayRate = etaDecayRate
            };
            SelfTune.ValidateTaskData(taskData);

            if (randomState == null)
            {
                this.random = new Random();
            }
            else
            {
                this.random = new Random(randomState.Value);
            }

            for (int i = 0; i < initialValues.Count; i++)
            {
                taskData.InitialValues[i] = SelfTune.Project(taskData.InitialValues[i], taskData.Constraints[i], taskData.Delta);
            }

            this.sessionData = SelfTune.GetNewSession(taskData.InitialValues, this.SampleUnitSphere(taskData.InitialValues.Count), taskData, 0);
            this.taskData = taskData;
        }

        /// <summary>
        /// Get the parameter values to use.
        /// Until <see cref="SetReward(double)"/> is called, calls to this will return the same value.
        /// </summary>
        /// <returns>The parameter values to use.</returns>
        public IList<double> Predict()
        {
            return this.sessionData.FirstReward == null ? this.sessionData.FirstExplore : this.sessionData.SecondExplore;
        }

        /// <summary>
        /// Notify the algorithm of how well the parameter values returned in <see cref="Predict"/> performed.
        /// </summary>
        /// <param name="reward">Measure of how well the parameter values returned in <see cref="Predict"/> performed.</param>
        public void SetReward(double reward)
        {
            if (this.taskData.Feedback == Feedback.OnePoint)
            {
                IList<double> newCenter = SelfTune.ComputeNewCenter(this.sessionData.Center, this.sessionData.ExploreDirection, reward, this.taskData);
                this.sessionData = SelfTune.GetNewSession(newCenter, this.SampleUnitSphere(newCenter.Count), this.taskData, this.sessionData.FirstExploreId + 1);
                this.round += 1;
            }
            else
            {
                if (!this.sessionData.FirstReward.HasValue)
                {
                    // reward for w + delta*u is now available
                    this.sessionData.FirstReward = reward;
                }
                else
                {
                    // reward for w - delta*u is now available, use this with the reward for w + delta*u to compute new center
                    IList<double> newCenter = SelfTune.ComputeNewCenter(this.sessionData.Center, this.sessionData.ExploreDirection, this.sessionData.FirstReward.Value - reward, this.taskData);
                    this.round += 1;
                    this.sessionData = SelfTune.GetNewSession(newCenter, this.SampleUnitSphere(newCenter.Count), this.taskData, this.sessionData.SecondExploreId + 1);
                }
            }
        }

        /// <summary>
        /// Sample uniformly from the surface of a unit sphere in n dimensions.
        /// </summary>
        private IList<double> SampleUnitSphere(int n)
        {
            List<double> sample = new List<double>(n);

            double norm = 0;
            for (int i = 0; i < n; i++)
            {
                double currentSample = this.SampleStandardNormal();
                norm += currentSample * currentSample;
                sample.Add(currentSample);
            }

            // norm == 0 is very unlikely, nevertheless handling it.
            if (norm == 0)
            {
                if (n > 0)
                {
                    sample[0] = (2 * this.random.Next(2)) - 1;
                }
            }
            else
            {
                norm = Math.Sqrt(norm);
                for (int i = 0; i < n; i++)
                {
                    sample[i] /= norm;
                }
            }

            return sample;
        }

        /// <summary>
        /// Sample from a standard normal distribution, i.e. ~ N(0, 1)
        /// </summary>
        private double SampleStandardNormal()
        {
            double u1 = 1 - this.random.NextDouble();
            double u2 = 1 - this.random.NextDouble();
            return Math.Sqrt(-2 * Math.Log(u1)) * Math.Sin(2 * Math.PI * u2);
        }

        /// <summary>
        /// Projects the center based on constraints.
        /// </summary>
        /// <param name="val">Value to project.</param>
        /// <param name="constraints">Constraints on the value.</param>
        /// <param name="delta">The exploration radius.</param>
        /// <returns>The projected set of parameters.</returns>
        private static double Project(double val, Constraints constraints, double delta)
        {
            double min = constraints.Min, max = constraints.Max;
            if (constraints.IsInt)
            {
                min = Math.Ceiling(min);
                max = Math.Floor(max);
                delta = Math.Ceiling(delta);
            }

            min += delta;
            max -= delta;

            val = Math.Min(Math.Max(min, val), max);
            if (constraints.IsInt)
            {
                val = Math.Round(val);
            }

            return val;
        }

        /// <summary>
        /// Decay eta based on the decay rate and current round.
        /// </summary>
        private double EtaUpdate()
        {
            return this.taskData.InitialEta / (1 + this.taskData.EtaDecayRate * this.round);
        }

        /// <summary>
        /// Compute the updated set of parameters using the reward(s) obtained during
        /// exploration around the current set of parameters.
        /// </summary>
        /// <param name="center">Current parameter vector.</param>
        /// <param name="exploreDirection">Direction of exploration.</param>
        /// <param name="rewardGradient">Contribution of the reward(s) towards the gradient.</param>
        /// <param name="taskData">Task for which new center is to be computed.</param>
        /// <returns>New set of parameters.</returns>
        private static IList<double> ComputeNewCenter(
            IList<double> center,
            IList<double> exploreDirection,
            double rewardGradient,
            TaskData taskData)
        {
            List<double> newCenter = new List<double>(center.Count);

            for (int i = 0; i < center.Count; i++)
            {
                newCenter.Add(
                    SelfTune.Project(
                        center[i] + (taskData.Eta * center.Count * rewardGradient * exploreDirection[i] / (((int)taskData.Feedback) * taskData.Delta)),
                        taskData.Constraints[i],
                        taskData.Delta));
            }

            return newCenter;
        }

        /// <summary>
        /// Compute the point to explore.
        /// </summary>
        /// <param name="center">The point around which exploration is done.</param>
        /// <param name="exploreSign">Whether to explore in the direction of exploreDirection or opposite to it.</param>
        /// <param name="exploreDirection">The direction along which to explore.</param>
        /// <param name="taskData">Task data.</param>
        /// <returns>Point to explore.</returns>
        private static IList<double> ComputeExploreValue(
            IList<double> center,
            ExploreSign exploreSign,
            IList<double> exploreDirection,
            TaskData taskData)
        {
            int numParameters = center.Count;
            List<double> exploreValue = new List<double>(numParameters);

            for (int i = 0; i < center.Count; i++)
            {
                double change = ((int)exploreSign) * taskData.Delta * exploreDirection[i];
                if (taskData.Constraints[i].IsInt)
                {
                    double rounded = Math.Round(change);
                    if (rounded == 0)
                    {
                        change = change >= 0 ? 1 : -1;
                    }
                    else
                    {
                        change = rounded;
                    }
                }
                exploreValue.Add(center[i] + change);
            }
            return exploreValue;
        }

        private static SessionData GetNewSession(
            IList<double> center,
            IList<double> exploreDirection,
            TaskData taskData,
            uint firstId)
        {
            return new SessionData
            {
                FirstExplore = SelfTune.ComputeExploreValue(center, ExploreSign.Positive, exploreDirection, taskData),
                SecondExplore = taskData.Feedback == Feedback.OnePoint ? null : SelfTune.ComputeExploreValue(center, ExploreSign.Negative, exploreDirection, taskData),
                Center = center,
                ExploreDirection = exploreDirection,
                FirstExploreId = firstId,
                SecondExploreId = taskData.Feedback == Feedback.OnePoint ? 0 : firstId + 1
            };
        }

        private static void ValidateTaskData(TaskData taskData)
        {
            int n = taskData.InitialValues.Count;

            if (taskData.InitialValues.Count != n)
            {
                throw new ArgumentException("Initial values not provided for all parameters");
            }

            if (taskData.Constraints.Count != n)
            {
                throw new ArgumentException("Constraints not provided for all parameters");
            }

            for (int i = 0; i < n; i++)
            {
                if (taskData.Constraints[i].Min > taskData.Constraints[i].Max)
                {
                    throw new ArgumentException($"Invalid constraint: Min(={taskData.Constraints[i].Min}) > Max({taskData.Constraints[i].Max})");
                }
            }

            if (taskData.Delta <= 0)
            {
                throw new ArgumentException($"Delta(={taskData.Delta}) should be > 0");
            }
        }
    }
}
